executor: decrease the memory usage of hashTable in HashJoinExec

SunRunAway opened the pull request:
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

part of #11607


### What is changed and how it works?

See  section 1 of https://github.com/pingcap/tidb/issues/11607#issuecomment-522400044

See benchmark at https://github.com/pingcap/tidb/pull/11832#issuecomment-523897237

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression

Related changes

 - None

XuHuaiyu commented:
Do we have any memory benchmark for this?

SunRunAway commented:
/run-all-tests

SunRunAway commented:
name             old time/op    new time/op    delta
HashJoinExec-12     894ms ±13%     167ms ± 4%   -81.30%  (p=0.000 n=9+10)

name             old alloc/op   new alloc/op   delta
HashJoinExec-12     853MB ± 0%     311MB ± 0%   -63.55%  (p=0.000 n=10+10)

name             old allocs/op  new allocs/op  delta
HashJoinExec-12      215k ± 0%      706k ± 0%  +228.89%  (p=0.000 n=10+10)

SunRunAway commented:
/run-all-tests

SunRunAway commented:
/run-all-tests

XuHuaiyu commented:
``` shell
executor/index_lookup_join.go
102:    lookupMap         *mvmap.MVMap
368:            lookupMap:         mvmap.NewMVMap(),

expression/aggregation/util.go
26:     existingKeys *mvmap.MVMap
35:             existingKeys: mvmap.NewMVMap(),

```

It seems that these two usages can be optimized, too. Maybe we can create an issue for it?

SunRunAway commented:
@XuHuaiyu 
Yes, these are the other tasks of issue https://github.com/pingcap/tidb/issues/11607

qw4990 commented on executor/join.go:
```diff
@@ -50,7 +50,7 @@ type HashJoinExec struct {
-	hashTable       *mvmap.MVMap
+	hashTable       rowHashTable
 	innerFinished   chan error
 	hashJoinBuffers []*hashJoinBuffer
```
It seems that this field can be removed.

XuHuaiyu commented on util/chunk/column.go:
```diff
@@ -605,3 +606,16 @@ func (c *Column) CopyReconstruct(sel []int, dst *Column) *Column {
+		elemLen := len(c.elemBuf)
+		data = c.data[rowID*elemLen : rowID*elemLen+elemLen]
+	} else {
+		start, end := c.offsets[rowID], c.offsets[rowID]
```
Does it seem to be wrong that `start == end`, any test case covered this?

SunRunAway replied:
The benchmark was fake because of this bug. 😢
I'll update a new one.

qw4990 commented on util/codec/codec.go:
```diff
@@ -284,105 +265,61 @@ func sizeInt(comparable bool) int {
-					return nil, errors.Trace(err)
+			flag := varintFlag
+			if mysql.HasUnsignedFlag(allTypes[i].Flag) {
+				if integer := row.GetInt64(i); integer < 0 {
```
```suggestion
				if row.GetInt64(i) < 0 {
```

qw4990 commented on util/codec/codec.go:
```diff
@@ -284,105 +265,61 @@ func sizeInt(comparable bool) int {
+			_, err = w.Write(bin)
 		default:
-			return nil, errors.Errorf("unsupport column type for encode %d", allTypes[i].Tp)
+			_, err = row.WriteTo(i, w)
```
It's dangerous to skip the type interpretation and hash the raw data directly.
For example:
```
create table t1(x float);
create table t2(x double);
insert into t1 values (1);
insert into t2 values (1);
select * from t1, t2 where t1.x=t2.x;
```

SunRunAway replied:
Since that, I've no good solution except changing all `encodeXXX` function along with an `io.Writer` argument.
I'll revert this part code for now and try to optimize it in further PR.
Any suggestion is very welcome, @qw4990 @zz-jason @XuHuaiyu

SunRunAway commented on executor/join.go:
```diff
@@ -390,7 +398,8 @@ func (e *HashJoinExec) runJoinWorker(workerID uint) {
 			break
 		}
-		ok, joinResult = e.join2Chunk(workerID, outerResult, joinResult, selected)
+		// TODO: a better deal with these three arguments: h, buf1, buf2
```
I'll address this TODO in next PR.

SunRunAway commented:
/run-all-tests

SunRunAway commented:
@XuHuaiyu @qw4990 
All comments are addressed, PTAL again, thanks.

zz-jason commented on util/codec/codec.go:
```diff
@@ -284,105 +266,144 @@ func sizeInt(comparable bool) int {
-	return encodeHashChunkRow(sc, b, row, allTypes, colIdx)
+// EqualChunkRow returns a boolean reporting whether row1 and row2
+// with their types and column index are logically equal.
+func EqualChunkRow(sc *stmtctx.StatementContext,
```
You can use `CompareFunc()` to do this, and it does not encode the compare fields. `CompareFunc` is located at `util/chunk/compare.go`

SunRunAway replied:
`CompareFunc` is good to compare two rows with the same field type in one table, but not good for comparing `uint64` and `int64` or comparing `float` and `double` in different table.

qw4990 replied:
How about using `Datum.CompareDatum`?
It should be faster than `EqualChunkRow` since it doesn't encode the data.

zz-jason replied:
It's better to do a benchmark to chose the best one.

SunRunAway replied:
`EqualChunkRow` does not encode the data either now.
I uses `chunk.GetCompareFunc` to benchmark with `EqualChunkRow`, the results are similar.

```
name                                                            old time/op    new time/op    delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     730ms ± 3%     732ms ± 2%   ~     (p=0.690 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12       165ms ± 4%     161ms ± 2%   ~     (p=0.151 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     305MB ± 0%     305MB ± 0%   ~     (p=0.151 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12       305MB ± 0%     305MB ± 0%   ~     (p=0.222 n=5+5)

name                                                            old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12      306k ± 0%      306k ± 0%   ~     (p=0.333 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12        306k ± 0%      306k ± 0%   ~     (p=0.421 n=5+5)
```

SunRunAway commented:
I've rewritten `HashChunkRow`, `EqualChunkRow` and `hashTable`, and the performance seems good.
The benchmark result is also updated.
@XuHuaiyu @qw4990 @zz-jason @wshwsh12 PTAL again, thanks.

SunRunAway commented:
/run-all-tests

qw4990 commented on executor/hash_table.go:
```diff
@@ -0,0 +1,99 @@
+type entryStore struct {
+	slices   [][]entry
+	sliceIdx uint32
+	sliceLen uint32
```
How about removing this field and replacing it with `len(slice[sliceIdx])`, which makes this structure simpler.

SunRunAway replied:
I just copied these codes from `mvmap.MVMap`, keeping `sliceLen` field make the code simple, and it's both ok for me.

qw4990 commented on executor/hash_table.go:
```diff
@@ -0,0 +1,99 @@
+
+var nullEntryAddr = entryAddr{}
+
+type rowHashMap struct {
```
It's not thread-safe, how about add some comments for it.

SunRunAway replied:
Done.

zz-jason replied:
Seems it's the same as `mvmap`. Can we reuse `mvmap`? the difference is, we use the encoded hash value as the key.

SunRunAway replied:
`mvmap` stores key into its `dataStore` and compare it each time when gets, while it's not necessary here.

qw4990 commented on executor/hash_table.go:
```diff
@@ -0,0 +1,99 @@
+		entryAddr = e.next
+		rowPtrs = append(rowPtrs, e.ptr)
+	}
+	// Keep the order of input.
```
Why do we do this?

SunRunAway replied:
I just copied these codes from `mvmap.MVMap`, I think it's good to keep the same behavior.

zz-jason commented on executor/join.go:
```diff
@@ -147,15 +141,9 @@ func (e *HashJoinExec) Open(ctx context.Context) error {
-		e.hashJoinBuffers = append(e.hashJoinBuffers, buffer)
+	e.joinKeyBuf = make([][]byte, e.concurrency)
+	for i := range e.joinKeyBuf {
+		e.joinKeyBuf[i] = make([]byte, 1)
```
why setting the size to `1`?

SunRunAway replied:
For memory reuse at https://github.com/pingcap/tidb/pull/11832/files#diff-671e79aa0fba47bd04f56fa4788120e2R360

lamxTyler commented on executor/hash_table.go:
```diff
@@ -0,0 +1,108 @@
+
+func (es *entryStore) put(e entry) entryAddr {
+	if es.sliceLen == maxEntrySliceLen {
+		es.slices = append(es.slices, make([]entry, 0, maxEntrySliceLen))
```
I am a little worried that this is too big for oltp cases, have you tested it?

SunRunAway replied:
I copied this line from `mvmap`, and I think it has a little problem.
In `newRowHashMap`, it makes a slice with 64 capacity but append a slice witch 8k capacity when first slice is full.

Since hash collision is not so frequent, it's not a big problem and I'll adjust it in further PR.

zz-jason replied:
How about other max capacities? For example, 32, 64, ...

zz-jason commented on executor/hash_table.go:
```diff
@@ -0,0 +1,108 @@
+		es.sliceIdx++
+	}
+	addr := entryAddr{sliceIdx: es.sliceIdx, offset: es.sliceLen}
+	slice := es.slices[es.sliceIdx]
```
```go
es.slices[es.sliceIdx] = append(es.slices[es.sliceIdx], e)
```
is much simpler.

SunRunAway commented:
@XuHuaiyu got two LGTMs, PTAL again~

XuHuaiyu commented on executor/hash_table.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// Put puts the key/rowPtr pairs to the rowHashMap, multiple rowPtrs are stored in a list.
+func (m *rowHashMap) Put(hashKey uint64, rowPtr chunk.RowPtr) {
```
If the origin rows are all stored in a Chunk but not a List, we need to pass `chunk.RowPtr{0, rowIdx}` into this function?

SunRunAway commented:
At least for now, but we can bench and optimize it later.

> On Aug 29, 2019, at 2:08 PM, HuaiyuXu <notifications@github.com> wrote:
> 
> @XuHuaiyu commented on this pull request.
> 
> In executor/hash_table.go <https://github.com/pingcap/tidb/pull/11832#discussion_r318898665>:
> 
> > +	length     int
> +}
> +
> +// newRowHashMap creates a new rowHashMap.
> +func newRowHashMap() *rowHashMap {
> +	m := new(rowHashMap)
> +	// TODO(fengliyuan): initialize the size of map from the estimated row count for better performance.
> +	m.hashTable = make(map[uint64]entryAddr)
> +	m.entryStore.slices = [][]entry{make([]entry, 0, 64)}
> +	// Reserve the first empty entry, so entryAddr{} can represent nullEntryAddr.
> +	m.entryStore.put(entry{})
> +	return m
> +}
> +
> +// Put puts the key/rowPtr pairs to the rowHashMap, multiple rowPtrs are stored in a list.
> +func (m *rowHashMap) Put(hashKey uint64, rowPtr chunk.RowPtr) {
> If the origin rows are all stored in a Chunk but not a List, we need to pass chunk.RowPtr{0, rowIdx} into this function?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pingcap/tidb/pull/11832?email_source=notifications&email_token=AAGTYNNX6DGJDQTERXDI4ETQG5RXDA5CNFSM4IOSJJNKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDBVAQQ#pullrequestreview-281235522>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAGTYNPU6HUPAERWWLDFAODQG5RXDANCNFSM4IOSJJNA>.
>

SunRunAway commented:
/run-common-test

SunRunAway commented:
/bench
